### PR TITLE
Do not get stuck due to adaptive capping

### LINKED ISSUE #748: Intensification may target the same configuration forever due to adaptive capping
#### Description

If a challenger run is skipped due to adaptive capping (adaptive cutoff drops below 0), the same challenger may be targeted in the next iteration of SMBO and be skipped for the same reason. This continues ad infinitum.

The configuration space includes a real-valued parameter, so a fresh configuration could be generated.
The target algorithm is considered deterministic.

#### Steps/Code to Reproduce

[infinite-loop.zip](https://github.com/automl/SMAC3/files/6925988/infinite-loop.zip)

- `out.log`: Beginning of the output of an infinite SMAC optimization. If I let it run longer, the last paragraph repeats forever.
- `run.sh`: Run this to repeat the SMAC optimization. Note that the target algorithm replays a run history, so it depends tightly on the SMAC version, seed, etc.

To reach the looping state in a debugger, set a breakpoint at `smac/intensification/intensification.py:813`.

#### Expected Results

A new challenger is evaluated eventually.

#### Actual Results

One challenger is evaluated repeatedly ad infinitum. See `out.log` in the attached archive.

#### Versions

SMAC: 0.14.0 3df3a749f3050d971af4110d051dae0cd795d615
Python: 3.8.10
OS: Ubuntu 20.04.2

### PR DESCRIPTION
**Pull request by filipbartek**
If the incumbent has all the instance-seed pairs covered
and the challenger is ruled out due to adaptive capping,
we cancel all the runs on the current challenger
to proceed to the next challenger.

This is a possibly dirty quick fix.
I am not well acquainted with the code.
I welcome suggestions for improvements of the fix.

Resolves #748

### CODE REVIEW

**Comment by renesass** [comment]
Thanks for your contribution! We'll look into this shortly.

**Comment by mlindauer** [comment]
Hi,

this looks good and the change makes sense in principle. However, I still don't fully see why the bug happens. 
@renesass could you please run the example from @filipbartek 's issue and check that everything is fine now.

@franchuterivera do you have any comment on this? I think this bug was introduced because of the new IntensifierStage, which I'm not super familiar with.

**Comment by renesass** [comment]
I ran @filipbartek code on both the old and the new code and got the same results. However, we should investigate why this problem happens in the first place. Since more time is required to find the bugs, we can't solve it by the upcoming release.

**Comment by mlindauer** [comment]
@franchuterivera @mfeurer could you please comment on this?
I don't understand why this change is required if we do a stage change anyway.
After we got not incumbent runs, I would expect that we could consider the same challenger again, since the adaptive capping allows now for more time.

**Comment by mlindauer** [comment]
@filipbartek please redirect your PR to `development`. Other than new releases, there is never a PR directed to `master`.

**Comment by mfeurer** [comment]
So if I understand this correctly, the function `_process_racer_results` does not know that the evaluation of the current challenger should be stopped because adaptive capping mandates that (as that functionality was simply not programmed) and therefore the stage change doesn't happen properly. So what you're doing right now @filipbartek  is "faking" the stop mechanism that `_process_racer_results` knows of. A cleaner solution would then be that we add a new argument to `_process_racer_results` which has the same effect as clearing `self.to_run` but is more descriptive. Is that correct @dengdifan?

Why the stage change wasn't sufficient is something I currently don't understand myself either :(

**Comment by dengdifan** [comment]
adding an argument to `_process_racer_results` could also be an option. 
Alternatively, we could add a docstring saying that "since the challenger fails to outperform the incumbent due to adaptive capping, we discard all the forthcoming runs"


> Why the stage change wasn't sufficient is something I currently don't understand myself either :(

Intensifier will skip the 4th line of `_process_racer_results` if `self.to_run` is not empty. As `self.to_run` still contains elements, intensifier will continue evaluating the remaining instance-seed pairs in `self.to_run`.

File: smac/intensification/intensification.py [hunk 1]
```diff
@@ -402,6 +402,11 @@ def get_next_run(self,
 
                 # If no more time, stage transition is a must
                 if not is_there_time_due_to_adaptive_cap:
+                    # The challenger is ruled out by adaptive capping.
+                    # We prevent this challenger from running again so that a new challenger is targeted in the next iteration.
+                    # Otherwise, the current challenger would be ruled out by adaptive capping again.
+                    self.continue_challenger = False
+                    self.to_run = []
```
[the remark this answers is not in the document]
Comment (filipbartek) [reply smac/intensification/intensification.py:410]: Having tested this on an example that I attached to issue #748, I agree that setting `self.to_run` suffices.
I have simplified the PR accordingly.

**Review by dengdifan — APPROVED** [review]

**Pull request merged**